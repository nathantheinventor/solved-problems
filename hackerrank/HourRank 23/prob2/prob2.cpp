#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <sstream>

#include <typeinfo>
#include <map>
using namespace std;

typedef vector<int> vec;

int convert(int x) {
    return x;
}

struct Node {
    Node* subNodes[10] = {nullptr};
    int elems = 0;
    int size;
    Node(int n) {
        size = n;
    }
    void insert(int x) {
        cout << "inserting " << x << endl;
        elems ++;
        if (size == 0) {
            return;
        }
        int sub = x / size;
        if (subNodes[sub] == nullptr) {
            subNodes[sub] = new Node(size / 10);
        }
        subNodes[sub]->insert(x % size);
    }
    int greater(int x) {
        if (size == 0) {
            return 0;
        }
        int ans = 0;
        int sub = x / size;
        for (int i = sub + 1; i < 10; i ++) {
            if (subNodes[i] != nullptr) {
                ans += subNodes[i]->elems;
            }
        }
        if (subNodes[sub] != nullptr) {
            ans += subNodes[sub]->greater(x % size);
        }
        return ans;
    }
};

int main() {
    Node tree(100000);
    int n, m, k;
    cin >> n >> m >> k;
    long long sum = 0;
    vector<int> nums(n, 0);
    map<int, vec* > indices;
    for (int i = 0; i < n; i ++) {
        int x;
        cin >> x;
        sum += x;
        nums[i] = x;
        if (indices.find(x) != indices.end()) {
            indices[x]->push_back(i + 1);
        } else {
            vector<int> *v = new vector<int>();
            v->push_back(i + 1);
            indices[x] = v;
        }
    }
    //vector<int> sorted_nums = nums;
    //sort(sorted_nums.begin(), sorted_nums.end());
    vector<int> s(k, 0);
    for (int i = 0; i < k; i ++) {
        int x;
        cin >> x;
        for (int j = 0; j < nums.size(); j ++) {
            if (nums[j] == x) {
                //cout << "inserting " << j + 1 << endl;
                tree.insert(j + 1);
            }
        }
        s[i] = x;
    }
    
    for (int i = 0; i < m; i ++) {
        long long a, b, c;
        cin >> a >> b >> c;
        sum += c * (b - a + 1);
        
        long long starred = tree.greater(a - 1) - tree.greater(b);
        //cout << starred << endl;
        sum -= starred * c;
        
        
        for (int j = 0; j < k; j ++) {
            s[j] -= c;
            if (indices.find(s[j]) != indices.end()) {
                vector<int> v = *indices[s[j]];
                vector<int> v2 = v;
                //cout << v.size() << endl;
                for (int e = 0; e < v.size(); e ++) {
                    stringstream ss;
                    int f;
                    ss << v.at(e);
                    ss >> f;

                    if (f <= b and f >= a and tree.greater(f) == tree.greater(f - 1)) {
                        tree.insert(f);
                    }
                }
            }
            
        }
        
        cout << sum << "\n";
    }
    return 0;
}
