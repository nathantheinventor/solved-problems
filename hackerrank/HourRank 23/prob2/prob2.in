100 100 2
21 11 12 16 8 6 5 10 37 12 8 39 19 23 30 29 29 30 12 29 14 13 11 11 40 32 7 26 7 27 8 4 19 5 9 35 5 3 15 12 25 32 12 12 16 25 26 19 11 37 7 32 33 13 30 5 3 14 6 34 27 1 4 29 5 21 32 26 26 24 39 2 38 30 13 35 29 9 19 27 24 37 30 21 26 21 35 32 5 10 9 12 11 34 12 35 14 25 7 11
75 10
11 97 1
47 63 1
23 84 1
36 48 1
45 72 1
18 24 1
27 27 1
14 79 1
25 88 1
31 65 1
5 21 1
17 67 1
43 83 1
7 46 1
27 63 1
26 71 1
22 63 1
8 51 1
35 85 1
8 32 1
11 86 1
37 72 1
24 73 1
10 69 1
49 64 1
16 100 1
48 85 1
39 78 1
33 50 1
36 49 1
18 21 1
34 80 1
7 90 1
40 88 1
25 73 1
5 49 1
50 98 1
35 100 1
40 41 1
43 68 1
28 38 1
17 82 1
43 79 1
41 53 1
37 88 1
50 77 1
30 67 1
11 21 1
27 35 1
7 13 1
25 36 1
43 46 1
16 23 1
26 67 1
27 38 1
6 49 1
46 46 1
35 98 1
33 68 1
5 40 1
45 50 1
40 51 1
27 50 1
13 62 1
38 73 1
19 51 1
32 76 1
5 68 1
14 83 1
3 39 1
47 83 1
20 54 1
19 91 1
14 100 1
34 52 1
16 84 1
19 80 1
21 78 1
20 89 1
22 51 1
29 47 1
48 50 1
20 100 1
35 69 1
31 66 1
10 87 1
48 80 1
24 51 1
7 80 1
18 98 1
40 66 1
48 81 1
34 58 1
27 61 1
31 60 1
26 81 1
14 21 1
46 52 1
20 98 1
9 69 1