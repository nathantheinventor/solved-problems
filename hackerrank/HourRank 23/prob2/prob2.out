inserting 8
inserting 8
inserting 8
inserting 8
inserting 8
inserting 8
inserting 0
inserting 90
inserting 90
inserting 90
inserting 90
inserting 90
inserting 0
inserting 0
inserting 35
inserting 35
inserting 35
inserting 35
inserting 35
inserting 5
inserting 0
inserting 78
inserting 78
inserting 78
inserting 78
inserting 78
inserting 8
inserting 0
inserting 91
inserting 91
inserting 91
inserting 91
inserting 91
inserting 1
inserting 0
2014
2031
inserting 27
inserting 27
inserting 27
inserting 27
inserting 27
inserting 7
inserting 0
inserting 29
inserting 29
inserting 29
inserting 29
inserting 29
inserting 9
inserting 0
inserting 51
inserting 51
inserting 51
inserting 51
inserting 51
inserting 1
inserting 0
2091
2104
inserting 56
inserting 56
inserting 56
inserting 56
inserting 56
inserting 6
inserting 0
inserting 65
inserting 65
inserting 65
inserting 65
inserting 65
inserting 5
inserting 0
2131
2138
2138
inserting 72
inserting 72
inserting 72
inserting 72
inserting 72
inserting 2
inserting 0
2197
inserting 62
inserting 62
inserting 62
inserting 62
inserting 62
inserting 2
inserting 0
2253
2283
2299
2343
2378
2414
2445
2484
2520
2559
2603
2625
2692
2723
2765
2818
2831
2905
2937
2971
2988
3002
3006
3046
3119
3162
inserting 25
inserting 25
inserting 25
inserting 25
inserting 25
inserting 5
inserting 0
3203
inserting 12
inserting 12
inserting 12
inserting 12
inserting 12
inserting 2
inserting 0
3243
inserting 73
inserting 73
inserting 73
inserting 73
inserting 73
inserting 3
inserting 0
3284
inserting 50
inserting 50
inserting 50
inserting 50
inserting 50
inserting 0
inserting 0
inserting 82
inserting 82
inserting 82
inserting 82
inserting 82
inserting 2
inserting 0
3340
3342
3363
3372
inserting 53
inserting 53
inserting 53
inserting 53
inserting 53
inserting 3
inserting 0
3425
inserting 52
inserting 52
inserting 52
inserting 52
inserting 52
inserting 2
inserting 0
inserting 67
inserting 67
inserting 67
inserting 67
inserting 67
inserting 7
inserting 0
3453
3462
inserting 55
inserting 55
inserting 55
inserting 55
inserting 55
inserting 5
inserting 0
inserting 74
inserting 74
inserting 74
inserting 74
inserting 74
inserting 4
inserting 0
inserting 83
inserting 83
inserting 83
inserting 83
inserting 83
inserting 3
inserting 0
3502
inserting 64
inserting 64
inserting 64
inserting 64
inserting 64
inserting 4
inserting 0
inserting 77
inserting 77
inserting 77
inserting 77
inserting 77
inserting 7
inserting 0
3518
3545
3555
inserting 28
inserting 28
inserting 28
inserting 28
inserting 28
inserting 8
inserting 0
3561
3566
3573
3577
3585
inserting 66
inserting 66
inserting 66
inserting 66
inserting 66
inserting 6
inserting 0
3613
3621
inserting 13
inserting 13
inserting 13
inserting 13
inserting 13
inserting 3
inserting 0
inserting 33
inserting 33
inserting 33
inserting 33
inserting 33
inserting 3
inserting 0
inserting 48
inserting 48
inserting 48
inserting 48
inserting 48
inserting 8
inserting 0
3658
3659
3701
inserting 45
inserting 45
inserting 45
inserting 45
inserting 45
inserting 5
inserting 0
3723
inserting 39
inserting 39
inserting 39
inserting 39
inserting 39
inserting 9
inserting 0
3750
3753
3761
inserting 40
inserting 40
inserting 40
inserting 40
inserting 40
inserting 0
inserting 0
inserting 43
inserting 43
inserting 43
inserting 43
inserting 43
inserting 3
inserting 0
inserting 44
inserting 44
inserting 44
inserting 44
inserting 44
inserting 4
inserting 0
3776
inserting 23
inserting 23
inserting 23
inserting 23
inserting 23
inserting 3
inserting 0
inserting 24
inserting 24
inserting 24
inserting 24
inserting 24
inserting 4
inserting 0
inserting 49
inserting 49
inserting 49
inserting 49
inserting 49
inserting 9
inserting 0
3806
3822
3838
3860
3895
inserting 59
inserting 59
inserting 59
inserting 59
inserting 59
inserting 9
inserting 0
3932
inserting 7
inserting 7
inserting 7
inserting 7
inserting 7
inserting 7
inserting 0
inserting 34
inserting 34
inserting 34
inserting 34
inserting 34
inserting 4
inserting 0
inserting 37
inserting 37
inserting 37
inserting 37
inserting 37
inserting 7
inserting 0
3957
inserting 63
inserting 63
inserting 63
inserting 63
inserting 63
inserting 3
inserting 0
3973
inserting 38
inserting 38
inserting 38
inserting 38
inserting 38
inserting 8
inserting 0
3987
4020
4067
4072
4103
4129
4151
4183
4193
4201
4201
4242
4254
4266
4304
4317
4326
4360
4401
4410
4424
4432
4445
4456
4479
4487
4489
4528
4556
