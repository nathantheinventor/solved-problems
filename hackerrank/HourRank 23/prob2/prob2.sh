g++ -std=c++11 prob2.cpp -o prob2
./prob2 < prob2.in > prob2.out
diff prob2.expected.out prob2.out