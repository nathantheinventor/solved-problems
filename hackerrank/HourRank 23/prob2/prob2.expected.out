2014
2031
2091
2104
2131
2138
2139
2199
2255
2282
2298
2337
2371
2401
2424
2455
2483
2516
2556
2575
2634
2660
2694
2738
2748
2814
2843
2874
2888
2900
2904
2938
3002
3042
3075
3110
3148
3200
3202
3221
3225
3274
3302
3314
3355
3375
3400
3410
3413
3418
3424
3428
3436
3463
3467
3500
3501
3550
3574
3598
3603
3613
3628
3664
3690
3713
3744
3787
3839
3862
3887
3906
3951
4010
4018
4062
4098
4128
4169
4183
4190
4191
4238
4249
4258
4302
4315
4324
4360
4401
4405
4418
4422
4427
4430
4446
4454
4454
4487
4508
